backend-rust/src/graphql_api.rs

```diff
@@ -743,13 +747,189 @@ LIMIT 30", // WHERE slot_time > (LOCALTIMESTAMP - $1::interval)
+        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: Option<i64>,
+        #[graphql(desc = "Returns the elements in the list that come after the specified cursor.")]
+        after: Option<String>,
+        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: Option<i64>,
+        #[graphql(desc = "Returns the elements in the list that come before the specified cursor.")]
+        before: Option<String>,
+    ) -> ApiResult<connection::Connection<String, Contract>> {
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+        let query = ConnectionQuery::<i64>::new(
+            first,
+            after,
+            last,
+            before,
+            config.contract_connection_limit,
+        )?;
+
+        let mut row_stream = sqlx::query!(
+            r#"
+SELECT * FROM (
+    SELECT
+        contracts.index as index,
+        sub_index,
+        module_reference,
+        name as contract_name,
+        contracts.amount,
+        blocks.slot_time as block_slot_time,
+        init_block_height as block_height,
+        transactions.hash as transaction_hash,
+        accounts.address as creator
+    FROM contracts
+    JOIN blocks ON init_block_height=blocks.height
+    JOIN transactions ON init_block_height=transactions.block_height AND init_transaction_index=transactions.index
+    JOIN accounts ON transactions.sender=accounts.index
+  WHERE contracts.index > $1 AND contracts.index < $2
+  ORDER BY
+     (CASE WHEN $4 THEN contracts.index END) DESC,
+     (CASE WHEN NOT $4 THEN contracts.index END) ASC
+  LIMIT $3
+) AS contract_data
+ORDER BY contract_data.index ASC"#,
+            query.from,
+            query.to,
+            query.limit,
+            query.desc
+        )
+        .fetch(pool);
+
+        let mut connection = connection::Connection::new(true, true);
+
+        while let Some(row) = row_stream.try_next().await? {
+            let contract_address_index = row.index.try_into()?;
+            let contract_address_sub_index = row.sub_index.try_into()?;
+
+            let snapshot = ContractSnapshot {
+                block_height: row.block_height,
+                contract_address_index,
+                contract_address_sub_index,
+                contract_name: row.contract_name,
+                module_reference: row.module_reference,
+                amount: row.amount,
+            };
+
+            let contract = Contract {
+                contract_address_index,
+                contract_address_sub_index,
+                contract_address: format!(
+                    "<{},{}>",
+                    contract_address_index, contract_address_sub_index
+                ),
+                creator: row.creator.into(),
+                block_height: row.block_height,
+                transaction_hash: row.transaction_hash,
+                block_slot_time: row.block_slot_time,
+                snapshot,
+            };
```

Comment:
One approach is to change the `Contract` struct to match exactly what we fetch from the database, and then use  `sqlx::query_as!(Contract, ...)`.
We then implement the conversion and GraphQL API using [`#[Object]`](https://async-graphql.github.io/async-graphql/en/define_complex_object.html) instead of `#[derive(SimpleObject)]`.
I have solved this for the `Block` and `Transaction` structs already.

Reply:
There might be a few code lines duplicated here but not sure if moving the whole conversion into `query_as!()` macro makes it much better (still duplicated logic).

Reply:
Not exactly what I meant, but let's keep it like this for now, I can make a small PR to show what I mean.